README.md
# azure-credentials

A tool to provide assistance for those struggling to create Applications, Service Principals and perform Role Assignments in Azure Resource Manager.
Generate your credentials file/helper for Chef, Puppet, Terraform and others!

## Background

Automated access to Azure Resource Manager usually requires PowerShell or the Azure-CLI and to follow the (many) manual steps in [create and authenticate a service principal](https://azure.microsoft.com/en-us/documentation/articles/resource-group-authenticate-service-principal/#authenticate-service-principal-with-password---azure-cli) to allow you to eventually retrieve 4 pieces of information:

1. **Subscription ID** 
2. **Client ID**
3. **Client Secret/Password**
4. **Tenant ID**

... you then use this information to configure your provisioning system. This tool (created in Ruby) requires only the credentials of the user with global administrator privilege and does the rest of the work for you, making it easier and quicker to get started with the various provisioning and testing toolsets out there.  

## Installation

The tool is distributed as a [Ruby Gem](https://rubygems.org/gems/azure-credentials).  To install it, run:

```gem install azure-credentials```

## Usage

```
Usage: azure-credentials (options)
    -l, --log_level LEVEL            Set the log level (debug, info, warn, error, fatal)
    -v, --verbose                    Display the credentials in STDOUT after creation? (warning: will contain secrets)
    -o, --output FILENAME            Enter the filename to save the credentials to
    -p, --password PASSWORD          Enter the password for the Azure AD user
    -r, --role ROLENAME              Enter the built-in Azure role to add the service principal to on your subscription (default: Contributor)
    -s, --subscription ID            Enter the Subscription ID to work against (default: process all subscriptions within the Azure tenant)
    -t, --type OUTPUTTYPE            Set the output type (default: chef)
    -u, --username USERNAME          Enter the username (must be an Azure AD user)
    -h, --help                       Show this message
```

## Default behaviour

By default (providing no options), the utility will prompt for your Azure AD username and password interactively, connect to Azure, retrieve details of all the subscriptions in your tenant, create an Application/Service Principal for each subscription then create you a credentials file with the relevant details that you can start to use with other tools immediately.

### Example output *(nb: ```chef exec``` typically is only required on Windows systems with ChefDK and no Ruby in their system path)*

```
PS C:\Users\StuartPreston\azure-credentials> chef exec azure-credentials
Enter your Azure AD username (user@domain.com): stuart@pendrica.com
Enter your password: 
INFO [2016-03-28 15:29:37] Authenticating to Azure Active Directory
INFO [2016-03-28 15:29:39] Retrieving tenant info
INFO [2016-03-28 15:29:40] Retrieving subscriptions info
INFO [2016-03-28 15:29:41] Creating application azure_6b3a_b6e7eee9-e4cd-41ae-8432-03ab624df016 in tenant 9c117323-1f20-444d-82a9-9ee430723ba3
INFO [2016-03-28 15:29:41] Creating service principal for application
INFO [2016-03-28 15:29:43] Waiting for service principal to be available in directory (retry 1)
INFO [2016-03-28 15:29:45] Attempting to assign service principal to role
INFO [2016-03-28 15:29:45] Waiting for service principal to be available in directory (retry 2)
INFO [2016-03-28 15:29:47] Attempting to assign service principal to role
INFO [2016-03-28 15:29:48] Waiting for service principal to be available in directory (retry 3)
INFO [2016-03-28 15:29:50] Attempting to assign service principal to role
INFO [2016-03-28 15:29:52] Assigned service principal to role Contributor in subscription b6e7eee9-e4cd-41ae-8432-03ab624df016
INFO [2016-03-28 15:29:52] Creating application azure_6b3a_bcf669fc-98b4-4e51-8ab0-e2d1f9f4b1c3 in tenant 9c117323-1f20-444d-82a9-9ee430723ba3
INFO [2016-03-28 15:29:53] Creating service principal for application
INFO [2016-03-28 15:29:54] Waiting for service principal to be available in directory (retry 1)
INFO [2016-03-28 15:29:56] Attempting to assign service principal to role
INFO [2016-03-28 15:29:56] Waiting for service principal to be available in directory (retry 2)
INFO [2016-03-28 15:29:58] Attempting to assign service principal to role
INFO [2016-03-28 15:29:58] Waiting for service principal to be available in directory (retry 3)
INFO [2016-03-28 15:30:00] Attempting to assign service principal to role
INFO [2016-03-28 15:30:04] Assigned service principal to role Contributor in subscription bcf669fc-98b4-4e51-8ab0-e2d1f9f4b1c3
INFO [2016-03-28 15:30:04] Creating credentials file at C:/Users/StuartPreston/azure-credentials/credentials
INFO [2016-03-28 15:30:04] Done!
```

You'll be able to see the output in the created credentials file (don't worry you can change the location using the -o switch)

```
PS C:\Users\StuartPreston\azure-credentials> cat ./credentials
[b6e7eee9-e4cd-41ae-8432-03ab624df016]
client_id = "5a8758f2-a2b2-4eab-8203-28c0840ccd7f"
client_secret = "ZSNdI5hQ9toavJNVgRjMjA=="
tenant_id = "9c117323-1f20-444d-82a9-9ee430723ba3"

[bcf669fc-98b4-4e51-8ab0-e2d1f9f4b1c3]
client_id = "2205cbcd-84dc-4b4a-9fcf-4a27d773a56d"
client_secret = "KboMQ86BvMGR6vO7_fUd1A=="
tenant_id = "9c117323-1f20-444d-82a9-9ee430723ba3"
```

(don't worry these accounts don't exist any more!)

If you look in the Azure Service Management portal and navigate to Active Directory > [directory name] > Applications (Show: Applications my company owns) you will see Applications created with the **azure_** prefix.  New identifiers are generated each time so there will be no overlap with existing accounts:

![config](images/configuration_asm.png)

You will also see the role assignment when looking in the Azure Resource Manager portal > Subscriptions, selecting one and viewing the users:

![config](images/configuration_arm.png)

## Some more advanced examples

### Chef quick config!

```azure-credentials -u username@domain.onmicrosoft.com -p "yourpassword" -o ~/.azure/credentials```

### Puppet-style output (note ```-v``` displays the file on screen after creation)

```azure-credentials -u username@domain.onmicrosoft.com -p "yourpassword" -o ./azure.conf -t puppet -v ```

```
PS C:\Users\StuartPreston\azure-credentials> chef exec azure-credentials -u stuart@pendrica.com -p "mypassword" -o ./azure.conf -t puppet -v
INFO [2016-03-28 16:01:58] Authenticating to Azure Active Directory
INFO [2016-03-28 16:02:00] Retrieving tenant info
INFO [2016-03-28 16:02:01] Retrieving subscriptions info
INFO [2016-03-28 16:02:01] Creating application azure_5291_b6e7eee9-e4cd-41ae-8432-03ab624df016 in tenant 9c117323-1f20-444d-82a9-9ee430723ba3
INFO [2016-03-28 16:02:02] Creating service principal for application
INFO [2016-03-28 16:02:03] Waiting for service principal to be available in directory (retry 1)
INFO [2016-03-28 16:02:05] Attempting to assign service principal to role
INFO [2016-03-28 16:02:05] Waiting for service principal to be available in directory (retry 2)
INFO [2016-03-28 16:02:07] Attempting to assign service principal to role
INFO [2016-03-28 16:02:08] Waiting for service principal to be available in directory (retry 3)
INFO [2016-03-28 16:02:10] Attempting to assign service principal to role
INFO [2016-03-28 16:02:10] Waiting for service principal to be available in directory (retry 4)
INFO [2016-03-28 16:02:12] Attempting to assign service principal to role
INFO [2016-03-28 16:02:16] Assigned service principal to role Contributor in subscription b6e7eee9-e4cd-41ae-8432-03ab624df016
INFO [2016-03-28 16:02:16] Creating application azure_5291_bcf669fc-98b4-4e51-8ab0-e2d1f9f4b1c3 in tenant 9c117323-1f20-444d-82a9-9ee430723ba3
INFO [2016-03-28 16:02:17] Creating service principal for application
INFO [2016-03-28 16:02:17] Waiting for service principal to be available in directory (retry 1)
INFO [2016-03-28 16:02:19] Attempting to assign service principal to role
INFO [2016-03-28 16:02:20] Waiting for service principal to be available in directory (retry 2)
INFO [2016-03-28 16:02:22] Attempting to assign service principal to role
INFO [2016-03-28 16:02:22] Waiting for service principal to be available in directory (retry 3)
INFO [2016-03-28 16:02:24] Attempting to assign service principal to role
INFO [2016-03-28 16:02:26] Assigned service principal to role Contributor in subscription bcf669fc-98b4-4e51-8ab0-e2d1f9f4b1c3
INFO [2016-03-28 16:02:26] Creating credentials file at C:/Users/StuartPreston/azure-credentials/azure.conf
azure: {
 subscription_id: "b6e7eee9-e4cd-41ae-8432-03ab624df016"
 tenant_id: "9c117323-1f20-444d-82a9-9ee430723ba3"
 client_id: "8dfab0fc-65c5-46d0-be1d-d5d301b3cbc9"
 client_secret: "m_nh57rEZpGe5GFL8-Iceg=="
}

azure: {
 subscription_id: "bcf669fc-98b4-4e51-8ab0-e2d1f9f4b1c3"
 tenant_id: "9c117323-1f20-444d-82a9-9ee430723ba3"
 client_id: "951eab1c-7e77-4931-bb68-15ca2d05e69e"
 client_secret: "vzl0haDXU51OukanIiRwJQ=="
}

INFO [2016-03-28 16:02:26] Done!
```

### Terraform-style output (note ```-v``` displays the file on screen after creation)

```azure-credentials -u username@domain.onmicrosoft.com -p "yourpassword" -o ./tf.creds -t terraform -v ```

```
PS C:\Users\StuartPreston\azure-credentials> chef exec azure-credentials -u stuart@pendrica.com -p "mypassword" -o ./tf.creds -t terraform -v
INFO [2016-03-28 16:04:32] Authenticating to Azure Active Directory
INFO [2016-03-28 16:04:34] Retrieving tenant info
INFO [2016-03-28 16:04:34] Retrieving subscriptions info
INFO [2016-03-28 16:04:34] Creating application azure_c346_b6e7eee9-e4cd-41ae-8432-03ab624df016 in tenant 9c117323-1f20-444d-82a9-9ee430723ba3
INFO [2016-03-28 16:04:35] Creating service principal for application
INFO [2016-03-28 16:04:36] Waiting for service principal to be available in directory (retry 1)
INFO [2016-03-28 16:04:38] Attempting to assign service principal to role
INFO [2016-03-28 16:04:42] Assigned service principal to role Contributor in subscription b6e7eee9-e4cd-41ae-8432-03ab624df016
INFO [2016-03-28 16:04:42] Creating application azure_c346_bcf669fc-98b4-4e51-8ab0-e2d1f9f4b1c3 in tenant 9c117323-1f20-444d-82a9-9ee430723ba3
INFO [2016-03-28 16:04:42] Creating service principal for application
INFO [2016-03-28 16:04:43] Waiting for service principal to be available in directory (retry 1)
INFO [2016-03-28 16:04:45] Attempting to assign service principal to role
INFO [2016-03-28 16:04:45] Waiting for service principal to be available in directory (retry 2)
INFO [2016-03-28 16:04:47] Attempting to assign service principal to role
INFO [2016-03-28 16:04:52] Assigned service principal to role Contributor in subscription bcf669fc-98b4-4e51-8ab0-e2d1f9f4b1c3
INFO [2016-03-28 16:04:52] Creating credentials file at C:/Users/StuartPreston/azure-credentials/tf.creds
provider "azurerm" {
  subscription_id = "b6e7eee9-e4cd-41ae-8432-03ab624df016"
  client_id       = "8a5b76c7-721f-4710-b491-8d657c9a7497"
  client_secret   = "rGRLfflNSFo7O4t2vBU7pg=="
  tenant_id       = "9c117323-1f20-444d-82a9-9ee430723ba3"
}

provider "azurerm" {
  subscription_id = "bcf669fc-98b4-4e51-8ab0-e2d1f9f4b1c3"
  client_id       = "0af58eee-25ce-467a-8e7c-05a35783385f"
  client_secret   = "UvB-HJPoiOpujufeJjp5Ag=="
  tenant_id       = "9c117323-1f20-444d-82a9-9ee430723ba3"
}

INFO [2016-03-28 16:04:52] Done!
```

## Not supported

1. Currently this tool does not support certificate-based authentication. 
2. This tool will not work if MFA (multi-factor authentication) is set to **required** for administration of the Azure Active Directory domain.
3. This tool does not work if you sign in with an ADFS-federated account.

## Contributing

I'm sure many improvements can be made to this tool! Contributions to the project are welcome via submitting Pull Requests.

1. Fork it ( https://github.com/pendrica/azure-credentials/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request


CHANGELOG.md
# azure-credentials changelog

## v0.1.5 (2019-11-29)
- Add support for Azure CLI format (@rdonkin)
- Fix dependencies for bundled installs (@rdonkin)

## v0.1.4 (2018-09-28)

- Adding bundler configuration, fix style issues (@stuartpreston)
- Require io/console module to avoid login failures (@blackknight36)

## v0.1.3 (2016-08-16)

- Upgrading dependencies (@stuartpreston)

## v0.1.1 (2016-03-31)

- Adding debug to authentication HTTP requests (@stuartpreston)


lib/azure/utility/credentials.rb
# frozen_string_literal: true

require 'io/console'
require 'net/http'
require 'uri'
require 'json'
require 'securerandom'
require 'time'
require 'logger'
require 'mixlib/cli'

module Azure
  module Utility
    #
    # Options
    #
    class Options
      include Mixlib::CLI

      option :username,
             short: '-u',
             long: '--username USERNAME',
             description: 'Enter the username (must be an Azure AD user)',
             required: false

      option :password,
             short: '-p',
             long: '--password PASSWORD',
             description: 'Enter the password for the Azure AD user',
             required: false

      option :subscription_id,
             short: '-s',
             long: '--subscription ID',
             description: 'Enter the Subscription ID to work against (default: process all subscriptions within the Azure tenant)',
             required: false,
             default: nil

      option :role,
             short: '-r',
             long: '--role ROLENAME',
             description: 'Enter the built-in Azure role to add the service principal to on your subscription (default: Contributor)',
             in: %w[Contributor Owner],
             default: 'Contributor',
             required: false

      option :type,
             short: '-t',
             long: '--type OUTPUTTYPE',
             description: 'Set the output type (default: chef)',
             in: %w[chef puppet terraform azurecli generic],
             required: false,
             default: 'chef'

      option :app_name,
             short: '-n',
             long: '--name APP_NAME',
             description: 'Name for new app registration used by service principal (default: azure_<random>_<subscription_id>)',
             default: :auto_name,
             required: false

      option :log_level,
             short: '-l',
             long: '--log_level LEVEL',
             description: 'Set the log level (debug, info, warn, error, fatal)',
             default: :info,
             required: false,
             in: %w[debug info warn error fatal],
             proc: proc { |l| l.to_sym }

      option :output_file,
             short: '-o',
             long: '--output FILENAME',
             description: 'Enter the filename to save the credentials to',
             default: './credentials',
             required: false

      option :out_to_screen,
             short: '-v',
             long: '--verbose',
             description: 'Display the credentials in STDOUT after creation? (warning: will contain secrets)',
             default: false,
             required: false

      option :help,
             short: '-h',
             long: '--help',
             description: 'Show this message',
             on: :tail,
             boolean: true,
             show_options: true,
             exit: 0
    end

    #
    # Logger
    #
    class CustomLogger
      def self.log
        if @logger.nil?
          cli = Options.new
          cli.parse_options
          @logger = Logger.new STDOUT
          @logger.level = logger_level_for(cli.config[:log_level])
          @logger.formatter = proc do |severity, datetime, _progname, msg|
            "#{severity} [#{datetime.strftime('%Y-%m-%d %H:%M:%S')}] #{msg}\n"
          end
        end
        @logger
      end

      def self.logger_level_for(sym)
        case sym
        when :debug
          Logger::DEBUG
        when :info
          Logger::INFO
        when :warn
          Logger::WARN
        when :error
          Logger::ERROR
        when :fatal
          Logger::FATAL
        end
      end
    end

    #
    # Credentials
    #
    class Credentials
      AZURE_SERVICE_PRINCIPAL = '1950a258-227b-4e31-a9cf-717495945fc2'
      CONFIG_PATH = "#{ENV['HOME']}/.azure/credentials"

      def initialize
        cli = Options.new
        cli.parse_options
        CustomLogger.log.debug "Command line options: #{cli.config.inspect}"

        username = cli.config[:username] || username_stdin
        password = cli.config[:password] || password_stdin

        # Get Bearer token for user and pass through to main method
        token = azure_authenticate(username, password)
        if token.nil?
          error_message = 'Unable to acquire token from Azure AD provider.'
          CustomLogger.log.error error_message
          raise error_message
        end
        created_credentials = create_all_objects(token, cli.config)
        CustomLogger.log.debug "Credential details: #{created_credentials.inspect}"
        create_file(created_credentials, cli.config)
        CustomLogger.log.info 'Done!'
      end

      def username_stdin
        print 'Enter your Azure AD username (user@domain.com): '
        STDIN.gets.chomp
      end

      def password_stdin
        print 'Enter your password: '
        STDIN.noecho(&:gets).chomp
      end

      def create_file(created_credentials, config)
        file_name = config[:output_file] || './credentials'
        file_name_expanded = File.expand_path(file_name)
        CustomLogger.log.info "Creating credentials file at #{file_name_expanded}"
        output = ''

        style = config[:type] || 'chef'
        case style
        when 'chef' # ref: https://github.com/pendrica/chef-provisioning-azurerm#configuration
          created_credentials.each do |s|
            subscription_template = <<~CHEFEOH
              [#{s[:subscription_id]}]
              client_id = "#{s[:client_id]}"
              client_secret = "#{s[:client_secret]}"
              tenant_id = "#{s[:tenant_id]}"

            CHEFEOH
            output += subscription_template
          end
        when 'terraform' # ref: https://www.terraform.io/docs/providers/azurerm/index.html
          created_credentials.each do |s|
            subscription_template = <<~TFEOH
              provider "azurerm" {
                subscription_id = "#{s[:subscription_id]}"
                client_id       = "#{s[:client_id]}"
                client_secret   = "#{s[:client_secret]}"
                tenant_id       = "#{s[:tenant_id]}"
              }

            TFEOH
            output += subscription_template
          end
        when 'puppet' # ref: https://github.com/puppetlabs/puppetlabs-azure#installing-the-azure-module
          created_credentials.each do |s|
            subscription_template = <<~PPEOH
              azure: {
                subscription_id: "#{s[:subscription_id]}"
                tenant_id: "#{s[:tenant_id]}"
                client_id: "#{s[:client_id]}"
                client_secret: "#{s[:client_secret]}"
              }

            PPEOH
            output += subscription_template
          end
        when 'azurecli'
          created_credentials.each do |s|
            subscription_template = <<~AZURECLIEOH
              [default]
              subscription_id=#{s[:subscription_id]}
              tenant=#{s[:tenant_id]}
              client_id=#{s[:client_id]}
              secret=#{s[:client_secret]}

            AZURECLIEOH
            output += subscription_template
          end
        else # generic credentials output
          created_credentials.each do |s|
            subscription_template = <<~GENERICEOH
              azure_subscription_id = "#{s[:subscription_id]}"
              azure_tenant_id = "#{s[:tenant_id]}"
              azure_client_id = "#{s[:client_id]}"
              azure_client_secret = "#{s[:client_secret]}"

            GENERICEOH
            output += subscription_template
          end
        end
        File.open(file_name_expanded, 'w') do |file|
          file.write(output)
        end
        puts output if config[:out_to_screen]
      end

      def create_all_objects(token, config)
        tenant_id = get_tenant_id(token).first['tenantId']
        subscriptions = Array(config[:subscription_id])
        subscriptions = get_subscriptions(token) if subscriptions.empty?
        identifier = SecureRandom.hex(2)
        credentials = []
        subscriptions.each do |subscription|
          if config[:app_name] == :auto_name then
            new_application_name = "azure_#{identifier}_#{subscription}"
          else
            new_application_name = config[:app_name]
          end

          if config[:type] == 'azurecli' then
            new_client_secret = SecureRandom.uuid
          else
            new_client_secret = SecureRandom.urlsafe_base64(16, true)
          end

          application_id = create_application(tenant_id, token, new_application_name, new_client_secret)['appId']
          service_principal_object_id = create_service_principal(tenant_id, token, application_id)['objectId']
          role_name = config[:role] || 'Contributor'
          role_definition_id = get_role_definition(subscription, token, role_name).first['id']
          success = false
          counter = 0
          until success || counter > 5
            counter += 1
            CustomLogger.log.info "Waiting for service principal to be available in directory (retry #{counter})"
            sleep 2
            assigned_role = assign_service_principal_to_role_id(subscription, token, service_principal_object_id, role_definition_id)
            success = true unless assigned_role['error']
          end
          raise 'Failed to assign Service Principal to Role' unless success
          CustomLogger.log.info "Assigned service principal to role #{role_name} in subscription #{subscription}"
          new_credentials = {}
          new_credentials[:subscription_id] = subscription
          new_credentials[:client_id] = application_id
          new_credentials[:client_secret] = new_client_secret
          new_credentials[:tenant_id] = tenant_id
          credentials.push(new_credentials)
        end
        credentials
      end

      def get_subscriptions(token)
        CustomLogger.log.info 'Retrieving subscriptions info'
        subscriptions = []
        subscriptions_call = azure_call(:get, 'https://management.azure.com/subscriptions?api-version=2015-01-01', nil, token)
        subscriptions_call['value'].each do |subscription|
          subscriptions.push subscription['subscriptionId']
        end
        CustomLogger.log.debug "SubscriptionIDs returned: #{subscriptions.inspect}"
        subscriptions
      end

      def get_tenant_id(token)
        CustomLogger.log.info 'Retrieving tenant info'
        tenants = azure_call(:get, 'https://management.azure.com/tenants?api-version=2015-01-01', nil, token)
        tenants['value']
      end

      def create_application(tenant_id, token, new_application_name, new_client_secret)
        CustomLogger.log.info "Creating application #{new_application_name} in tenant #{tenant_id}"
        url = "https://graph.windows.net/#{tenant_id}/applications?api-version=1.42-previewInternal"
        payload_json = <<-JSONEOH
        {
            "availableToOtherTenants": false,
            "displayName": "#{new_application_name}",
            "homepage": "https://management.core.windows.net",
            "identifierUris": [
                "https://#{tenant_id}/#{new_application_name}"
            ],
            "passwordCredentials": [
                {
                "startDate": "#{Time.now.utc.iso8601}",
                "endDate": "#{(Time.now + (24 * 60 * 60 * 365 * 10)).utc.iso8601}",
                "keyId": "#{SecureRandom.uuid}",
                "value": "#{new_client_secret}"
                }
            ]
        }
        JSONEOH
        azure_call(:post, url, payload_json, token)
      end

      def create_service_principal(tenant_id, token, application_id)
        CustomLogger.log.info 'Creating service principal for application'
        url = "https://graph.windows.net/#{tenant_id}/servicePrincipals?api-version=1.42-previewInternal"
        payload_json = <<-PAYLOADEOH
        {
            "appId": "#{application_id}",
            "accountEnabled": true
        }
        PAYLOADEOH
        azure_call(:post, url, payload_json, token)
      end

      def assign_service_principal_to_role_id(subscription_id, token, service_principal_object_id, role_definition_id)
        CustomLogger.log.info 'Attempting to assign service principal to role'
        url = "https://management.azure.com/subscriptions/#{subscription_id}/providers/Microsoft.Authorization/roleAssignments/#{service_principal_object_id}?api-version=2015-07-01"
        payload_json = <<-PAYLOADEOH
        {
            "properties": {
                "roleDefinitionId": "#{role_definition_id}",
                "principalId": "#{service_principal_object_id}"
            }
        }
        PAYLOADEOH
        azure_call(:put, url, payload_json, token)
      end

      def get_role_definition(tenant_id, token, role_name)
        role_definitions = azure_call(:get, "https://management.azure.com/subscriptions/#{tenant_id}/providers/Microsoft.Authorization/roleDefinitions?$filter=roleName%20eq%20\'#{role_name}\'&api-version=2015-07-01", nil, token)
        role_definitions['value']
      end

      def azure_authenticate(username, password)
        CustomLogger.log.info 'Authenticating to Azure Active Directory'
        url = 'https://login.windows.net/Common/oauth2/token'
        data = "resource=https%3A%2F%2Fmanagement.core.windows.net%2F&client_id=#{AZURE_SERVICE_PRINCIPAL}" \
          "&grant_type=password&username=#{username}&scope=openid&password=#{password}"
        response = http_post(url, data)
        JSON.parse(response.body)['access_token']
      end

      def http_post(url, data)
        uri = URI(url)
        response = nil
        Net::HTTP.start(uri.host, uri.port, use_ssl: true) do |http|
          request = Net::HTTP::Post.new uri
          CustomLogger.log.debug "Request: #{request.uri} (#{request.method}) #{data}"
          request.body = data
          response = http.request request
          CustomLogger.log.debug "Response: #{response.body}"
        end
        response
      end

      def azure_call(method, url, data, token)
        uri = URI(url)
        response = nil
        Net::HTTP.start(uri.host, uri.port, use_ssl: true) do |http|
          case method
          when :put
            request = Net::HTTP::Put.new uri
          when :delete
            request = Net::HTTP::Delete.new uri
          when :get
            request = Net::HTTP::Get.new uri
          when :post
            request = Net::HTTP::Post.new uri
          when :patch
            request = Net::HTTP::Patch.new uri
          end
          request.body = data
          request['Authorization'] = "Bearer #{token}"
          request['Content-Type'] = 'application/json'
          CustomLogger.log.debug "Request: #{request.uri} (#{method}) #{data}"
          response = http.request request
          CustomLogger.log.debug "Response: #{response.body}"
        end
        JSON.parse(response.body) unless response.nil?
      end
    end
  end
end
